File: family-history-intake-form/node/src/intakeService.ts
/**
 * Orchestrates the sequence of Evagene REST calls that turn a validated
 * IntakeSubmission into a persisted pedigree with relatives wired up.
 *
 * No HTTP knowledge -- only the EvageneClient interface.  The order of
 * calls is load-bearing: grandparents depend on parents existing first
 * (the `relative_of` field points at an already-created individual).
 */

import type { EvageneApi, RelativeType } from './evageneClient.js';
import type {
  IntakeSubmission,
  RelativeEntry,
  SiblingEntry,
} from './intakeSubmission.js';

export interface IntakeServiceOptions {
  readonly client: EvageneApi;
}

export interface IntakeCreationResult {
  readonly pedigreeId: string;
  readonly probandId: string;
  readonly relativesAdded: number;
}

export class IntakeService {
  constructor(private readonly options: IntakeServiceOptions) {}

  async create(submission: IntakeSubmission): Promise<IntakeCreationResult> {
    const { client } = this.options;

    const pedigreeId = await client.createPedigree({
      displayName: `${submission.proband.displayName}'s family`,
    });
    const probandId = await client.createIndividual({
      displayName: submission.proband.displayName,
      biologicalSex: submission.proband.biologicalSex,
      ...(submission.proband.yearOfBirth !== undefined
        ? { yearOfBirth: submission.proband.yearOfBirth }
        : {}),
    });
    await client.addIndividualToPedigree(pedigreeId, probandId);
    await client.designateAsProband(probandId);

    const motherId = await this.maybeAddRelative(
      pedigreeId,
      probandId,
      'mother',
      'female',
      submission.mother,
    );
    const fatherId = await this.maybeAddRelative(
      pedigreeId,
      probandId,
      'father',
      'male',
      submission.father,
    );

    let relativesAdded = countAdded(motherId, fatherId);

    if (motherId !== undefined) {
      relativesAdded += await this.addGrandparents(
        pedigreeId,
        motherId,
        submission.maternalGrandmother,
        submission.maternalGrandfather,
        'maternal',
      );
    }
    if (fatherId !== undefined) {
      relativesAdded += await this.addGrandparents(
        pedigreeId,
        fatherId,
        submission.paternalGrandmother,
        submission.paternalGrandfather,
        'paternal',
      );
    }
    relativesAdded += await this.addSiblings(pedigreeId, probandId, submission.siblings);

    return { pedigreeId, probandId, relativesAdded };
  }

  private async addGrandparents(
    pedigreeId: string,
    parentId: string,
    grandmother: RelativeEntry | undefined,
    grandfather: RelativeEntry | undefined,
    side: 'maternal' | 'paternal',
  ): Promise<number> {
    const grandmotherId = await this.maybeAddRelative(
      pedigreeId,
      parentId,
      'mother',
      'female',
      grandmother,
    );
    const grandfatherId = await this.maybeAddRelative(
      pedigreeId,
      parentId,
      'father',
      'male',
      grandfather,
    );
    // side is carried only for future disambiguation / logging hooks.
    void side;
    return countAdded(grandmotherId, grandfatherId);
  }

  private async addSiblings(
    pedigreeId: string,
    probandId: string,
    siblings: readonly SiblingEntry[],
  ): Promise<number> {
    let added = 0;
    for (const sibling of siblings) {
      await this.options.client.addRelative({
        pedigreeId,
        relativeOf: probandId,
        relativeType: sibling.relation,
        displayName: sibling.displayName,
        biologicalSex: sibling.biologicalSex,
        ...(sibling.yearOfBirth !== undefined ? { yearOfBirth: sibling.yearOfBirth } : {}),
      });
      added += 1;
    }
    return added;
  }

  private async maybeAddRelative(
    pedigreeId: string,
    relativeOf: string,
    relativeType: RelativeType,
    biologicalSex: 'female' | 'male',
    entry: RelativeEntry | undefined,
  ): Promise<string | undefined> {
    if (entry === undefined) {
      return undefined;
    }
    return this.options.client.addRelative({
      pedigreeId,
      relativeOf,
      relativeType,
      displayName: entry.displayName,
      biologicalSex,
      ...(entry.yearOfBirth !== undefined ? { yearOfBirth: entry.yearOfBirth } : {}),
    });
  }
}

function countAdded(...ids: (string | undefined)[]): number {
  return ids.filter(id => id !== undefined).length;
}

File: family-history-intake-form/dotnet/src/FamilyIntake/IntakeService.cs
namespace FamilyIntake;

/// <summary>
/// Orchestrates the sequence of Evagene REST calls that turn a validated
/// IntakeSubmission into a persisted pedigree with relatives wired up.
///
/// No HTTP knowledge -- only the IEvageneApi interface. The order of
/// calls is load-bearing: grandparents depend on parents existing first
/// (the relative_of field points at an already-created individual).
/// </summary>
public sealed record IntakeCreationResult(string PedigreeId, string ProbandId, int RelativesAdded);

public sealed class IntakeService
{
    private readonly IEvageneApi client;

    public IntakeService(IEvageneApi client)
    {
        this.client = client;
    }

    public async Task<IntakeCreationResult> CreateAsync(
        IntakeSubmission submission,
        CancellationToken cancellationToken = default)
    {
        var pedigreeId = await this.client.CreatePedigreeAsync(
            new CreatePedigreeArgs($"{submission.Proband.DisplayName}'s family"),
            cancellationToken).ConfigureAwait(false);
        var probandId = await this.client.CreateIndividualAsync(
            new CreateIndividualArgs(
                submission.Proband.DisplayName,
                submission.Proband.BiologicalSex,
                submission.Proband.YearOfBirth),
            cancellationToken).ConfigureAwait(false);
        await this.client.AddIndividualToPedigreeAsync(pedigreeId, probandId, cancellationToken).ConfigureAwait(false);
        await this.client.DesignateAsProbandAsync(probandId, cancellationToken).ConfigureAwait(false);

        var motherId = await this.MaybeAddRelativeAsync(
            pedigreeId, probandId, RelativeType.Mother, BiologicalSex.Female,
            submission.Mother, cancellationToken).ConfigureAwait(false);
        var fatherId = await this.MaybeAddRelativeAsync(
            pedigreeId, probandId, RelativeType.Father, BiologicalSex.Male,
            submission.Father, cancellationToken).ConfigureAwait(false);

        var relativesAdded = CountAdded(motherId, fatherId);

        if (motherId is not null)
        {
            relativesAdded += await this.AddGrandparentsAsync(
                pedigreeId, motherId,
                submission.MaternalGrandmother, submission.MaternalGrandfather,
                cancellationToken).ConfigureAwait(false);
        }
        if (fatherId is not null)
        {
            relativesAdded += await this.AddGrandparentsAsync(
                pedigreeId, fatherId,
                submission.PaternalGrandmother, submission.PaternalGrandfather,
                cancellationToken).ConfigureAwait(false);
        }
        relativesAdded += await this.AddSiblingsAsync(
            pedigreeId, probandId, submission.Siblings, cancellationToken).ConfigureAwait(false);

        return new IntakeCreationResult(pedigreeId, probandId, relativesAdded);
    }

    private async Task<int> AddGrandparentsAsync(
        string pedigreeId,
        string parentId,
        RelativeEntry? grandmother,
        RelativeEntry? grandfather,
        CancellationToken cancellationToken)
    {
        var grandmotherId = await this.MaybeAddRelativeAsync(
            pedigreeId, parentId, RelativeType.Mother, BiologicalSex.Female,
            grandmother, cancellationToken).ConfigureAwait(false);
        var grandfatherId = await this.MaybeAddRelativeAsync(
            pedigreeId, parentId, RelativeType.Father, BiologicalSex.Male,
            grandfather, cancellationToken).ConfigureAwait(false);
        return CountAdded(grandmotherId, grandfatherId);
    }

    private async Task<int> AddSiblingsAsync(
        string pedigreeId,
        string probandId,
        IReadOnlyList<SiblingEntry> siblings,
        CancellationToken cancellationToken)
    {
        var added = 0;
        foreach (var sibling in siblings)
        {
            await this.client.AddRelativeAsync(
                new AddRelativeArgs(
                    pedigreeId,
                    probandId,
                    RelativeTypeForSibling(sibling.Relation),
                    sibling.DisplayName,
                    sibling.BiologicalSex,
                    sibling.YearOfBirth),
                cancellationToken).ConfigureAwait(false);
            added += 1;
        }
        return added;
    }

    private async Task<string?> MaybeAddRelativeAsync(
        string pedigreeId,
        string relativeOf,
        RelativeType relativeType,
        BiologicalSex biologicalSex,
        RelativeEntry? entry,
        CancellationToken cancellationToken)
    {
        if (entry is null)
        {
            return null;
        }
        return await this.client.AddRelativeAsync(
            new AddRelativeArgs(
                pedigreeId,
                relativeOf,
                relativeType,
                entry.DisplayName,
                biologicalSex,
                entry.YearOfBirth),
            cancellationToken).ConfigureAwait(false);
    }

    private static RelativeType RelativeTypeForSibling(SiblingRelation relation) => relation switch
    {
        SiblingRelation.Sister => RelativeType.Sister,
        SiblingRelation.Brother => RelativeType.Brother,
        SiblingRelation.HalfSister => RelativeType.HalfSister,
        SiblingRelation.HalfBrother => RelativeType.HalfBrother,
        _ => throw new ArgumentOutOfRangeException(nameof(relation)),
    };

    private static int CountAdded(params string?[] ids)
    {
        var count = 0;
        foreach (var id in ids)
        {
            if (id is not null)
            {
                count += 1;
            }
        }
        return count;
    }
}
